### DISCUSSION TITLE
FastAPI+Uvicorn Startup Event Handling Long-Running Functions

### DISCUSSION THREAD

**Discussion by Well2333** [encode/starlette#2322] [Q&A]
I'm using FastAPI with Uvicorn, and I'm encountering an issue with long-running asynchronous functions in the startup event (`@app.on_event("startup")`). These functions prevent the program from being interrupted gracefully (e.g., via Ctrl+C), which is critical for my use case where the startup phase includes potentially lengthy operations like setting up database connections or external API calls. This behavior might be related to FastAPI's underlying framework, starlette.

```python
import asyncio
from fastapi import FastAPI
import uvicorn

app = FastAPI()

@app.on_event("startup")
async def startup_event():
    print("startup")
    await asyncio.sleep(5)  # Simulating a long-running task
    print("startup complete")

if __name__ == "__main__":
    uvicorn.run(app, host="127.0.0.1", port=8000)
```

I would like the application to remain responsive to interrupt signals even during the execution of long-running startup tasks. Ideally, it should be possible to stop the server gracefully at any point, including during the startup phase.

I am looking for suggestions or solutions that could help in making the startup_event in FastAPI (possibly influenced by starlette) interruptible. This might include insights into starlette's handling of asynchronous tasks, custom signal handling, event loop management, or any architectural changes that can improve responsiveness to interrupts during startup.

**Comment by MrTONYCHAN** [comment]
ASGI lifespan startup is intentionally a readiness barrier: the server does not begin serving until startup completes. That makes a long operation in startup the wrong place for work that should be independently cancellable or allowed to continue after readiness.

Use a lifespan context and retain the task so it can be cancelled during shutdown:

```python
from contextlib import asynccontextmanager
import asyncio
from fastapi import FastAPI

async def background_initialization():
    await asyncio.sleep(30)

@asynccontextmanager
async def lifespan(app: FastAPI):
    task = asyncio.create_task(background_initialization())
    try:
        yield
    finally:
        task.cancel()
        await asyncio.gather(task, return_exceptions=True)

app = FastAPI(lifespan=lifespan)
```

Only use that pattern if the application can safely serve before initialization finishes. Expose readiness as false until required state is available.

If initialization must finish before any request is served, keep it in lifespan startup but make each external operation bounded with timeouts and cancellation-friendly async libraries, for example `asyncio.timeout(...)` on supported Python versions. Avoid swallowing `CancelledError`.

For database connections, initialize a pool rather than running migrations or other unbounded jobs in every web worker. Long migrations/data preparation are usually better as a separate deployment job before the service starts.
